engine/src/main/java/org/pentaho/di/pkg/JarfileGenerator.java

```diff
@@ -0,0 +1,188 @@
+/*! ******************************************************************************
+ *
+ * Pentaho Data Integration
+ *
+ * Copyright (C) 2002-2017 by Hitachi Vantara : http://www.pentaho.com
+ *
+ *******************************************************************************
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ ******************************************************************************/
+
+package org.pentaho.di.pkg;
+
+import java.io.File;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.jar.Attributes;
+
+import org.pentaho.di.core.Const;
+import org.pentaho.di.core.util.Utils;
+import org.pentaho.di.core.logging.LogChannel;
+import org.pentaho.di.core.logging.LogChannelInterface;
+import org.pentaho.di.core.util.EnvUtil;
+import org.pentaho.di.core.util.StreamLogger;
+import org.pentaho.di.core.xml.XMLHandler;
+import org.pentaho.di.trans.TransMeta;
+import org.pentaho.di.version.BuildVersion;
+
+public class JarfileGenerator {
+  private static LogChannelInterface log;
+
+  public static final String TRANSFORMATION_FILENAME = "transformation.xml";
+
+  public static final void generateJarFile( TransMeta transMeta ) {
+    log = new LogChannel( "Jar file generator" );
+    KettleDependencies deps = new KettleDependencies( transMeta );
+
+    File kar = new File( "kar" );
+    if ( kar.exists() ) {
+      log.logBasic( "Jar generator", "Removing directory: " + kar.getPath() );
+      deleteDirectory( kar );
+    }
+    kar.mkdir();
+
+    String filename = "kettle-engine-3.0.jar";
+    if ( !Utils.isEmpty( transMeta.getFilename() ) ) {
+      filename = Const.replace( transMeta.getFilename(), " ", "_" ).toLowerCase() + ".kar";
+    }
+
+    File karFile = new File( filename );
+
+    try {
+      // The manifest file
+      String strManifest = "";
+      strManifest += "Manifest-Version: 1.0" + Const.CR;
+      strManifest += "Created-By: Kettle version " + BuildVersion.getInstance().getVersion() + Const.CR;
+      strManifest += Attributes.Name.MAIN_CLASS.toString() + ": " + ( JarPan.class.getName() ) + Const.CR;
+
+      // Create a new manifest file in the root.
+      File manifestFile = new File( kar.getPath() + "/" + "manifest.mf" );
+      FileOutputStream fos = new FileOutputStream( manifestFile );
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *fos*. The resource is referenced at line 75. The resource is closed at line 76. There are other execution paths that do not contain closure statements, for example, when FileOutputStream.write throws an exception. Ensure *fos* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).